### PR TITLE
Time actions

### PR DESCRIPTION
**Pull request by wthrowe**

### CODE REVIEW

File: tests/Unit/Parallel/Test_AlgorithmNodelock.cpp [hunk 1]
```diff
@@ -61,10 +61,13 @@ struct nodegroup_receive {
     db::mutate<Tags::vector_of_array_indexs, Tags::total_receives_on_node>(
         box, [&id_of_array](std::vector<int>& array_indexs,
                             int& total_receives_on_node) {
-          if (array_indexs.size() !=
-              number_of_1d_array_elements_per_core *
-                  static_cast<size_t>(
-                      Parallel::procs_on_node(Parallel::my_node()))) {
+          // This, and the similarly names variables below, cause
```
Comment (kidder) [inline tests/Unit/Parallel/Test_AlgorithmNodelock.cpp]: names -> named

File: src/Time/TimeTags.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Distributed under the MIT License.
```
Comment (kidder) [inline src/Time/TimeTags.hpp]: move to Time/Tags.hpp

File: src/Time/TimeTags.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include "Options/Options.hpp"
+#include "Time/Time.hpp"
+#include "Time/TimeId.hpp"
+#include "Utilities/TMPL.hpp"
+
+class TimeStepper;
+
+namespace Tags {
+
+/// \ingroup DataBoxTags
```
Comment (kidder) [inline src/Time/TimeTags.hpp]: also (if it works) \ingroup Time

File: src/Time/Actions/AdvanceTime.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#include "DataStructures/DataBox.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Time/Time.hpp"
+#include "Time/TimeId.hpp"
+#include "Time/TimeSteppers/TimeStepper.hpp"
+#include "Time/TimeTags.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Actions {
+/// \ingroup Actions
```
Comment (kidder) [inline src/Time/Actions/AdvanceTime.hpp]: ActionsGroup

File: tests/Unit/ActionTesting.hpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  using index = Index;
+  using const_global_cache_tag_list = ConstGlobalCacheTagList;
+  using inbox_tag_list = InboxTagList;
+  using action_list = ActionList;
+  using initial_databox = int;  // type irrelevant
+};
+
+/// A class that mocks the infrastructure needed to run actions.  It
+/// simulates message passing using the inbox infrastructure and
+/// handles most of the arguments to the apply and is_readu action
```
Comment (kidder) [inline tests/Unit/ActionTesting.hpp]: ready

Comment (kidder) [reply tests/Unit/ActionTesting.hpp]: \ingroup TestingSomething

File: tests/Unit/ActionTesting.hpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      if (not tuples::get<InboxTag>(element_box.second).empty()) {
+        result.insert(element_box.first);
+      }
+    }
+    return result;
+  }
+
+ private:
+  template <typename T>
+  const T& as_const(const T& t) noexcept { return t; }
```
Comment (kidder) [inline tests/Unit/ActionTesting.hpp]: put into ConstantExpressions

File: tests/Unit/ActionTesting.hpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+          typename InboxTagList = tmpl::list<>,
+          typename ActionList = tmpl::list<>>
+struct MockArrayComponent {
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting_detail::MockArrayChare;
+  using index = Index;
+  using const_global_cache_tag_list = ConstGlobalCacheTagList;
+  using inbox_tag_list = InboxTagList;
+  using action_list = ActionList;
+  using initial_databox = int;  // type irrelevant
```
Comment (kidder) [inline tests/Unit/ActionTesting.hpp]: int -> NoSuchType

File: src/Time/Actions/UpdateU.hpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+/// \ingroup TimeGroup
+/// \brief Perform variable updates for one substep
+///
+/// Uses:
+/// - ConstGlobalCache: CacheTags::TimeStepper
+/// - DataBox: system::variables_tag, system::dt_variables_tag,
+///   Tags::HistoryEvolvedVariables<system::variables_tag,
+///                                 system::dt_variables_tag>,
+///   Tags::Time, Tags::TimeStep
+///
```
Comment (kidder) [inline src/Time/Actions/UpdateU.hpp:31]: Adds to DataBox:
Removes to DataBox:

File: src/Time/Actions/AdvanceTime.hpp [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+/// \ingroup TimeGroup
+/// \brief Advance time one substep
+///
+/// Uses:
+/// - ConstGlobalCache: CacheTags::TimeStepper
+/// - DataBox: Tags::TimeId, Tags::TimeStep
+///
+/// DataBox type is unmodified.
```
Comment (kidder) [inline src/Time/Actions/AdvanceTime.hpp]: Adds to DataBox:
Removes from DataBox:
Modifies in DataBox:

File: src/Time/TimeTags.hpp [hunk 10]
```diff
@@ -44,6 +46,28 @@ struct Time : db::ComputeItemTag {
   using argument_tags = tmpl::list<TimeId>;
 };
 
+/// \ingroup DataBoxTags
+/// \brief Prefix indicating a time derivative
+template <typename Tag>
+struct dt : db::DataBoxPrefix {
```
Comment (kidder) [inline src/Time/TimeTags.hpp]: move to DataBox/Prefixes.hpp

File: src/Time/Actions/FinalTime.hpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  static auto apply(db::DataBox<DbTags>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/) noexcept {
+    const double final_time = Parallel::get<CacheTags::FinalTime>(cache);
+    const Time& time = db::get<Tags::Time>(box);
+    const TimeDelta& time_step = db::get<Tags::TimeStep>(box);
+
+    return std::make_tuple(
```
Comment (kidder) [inline src/Time/Actions/FinalTime.hpp]: std::tuple<db::DataBox<DbTags>&&, bool> (std::move....

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/Parallel/Test_AlgorithmNodelock.cpp [hunk 12]
```diff
@@ -61,13 +61,9 @@ struct nodegroup_receive {
...
         box, [&id_of_array](std::vector<int>& array_indexs,
                             int& total_receives_on_node) {
-          // This, and the similarly names variables below, cause
-          // conversion warnings in gcc 6.3 when inlined into the
-          // following expressions.
-          const auto cant_inline_in_gcc_63 = static_cast<size_t>(
-                  number_of_1d_array_elements_per_core *
-                  Parallel::procs_on_node(Parallel::my_node()));
-          if (array_indexs.size() != cant_inline_in_gcc_63) {
+          if (static_cast<int>(array_indexs.size()) !=
```
Comment (kidder) [inline tests/Unit/Parallel/Test_AlgorithmNodelock.cpp]: :+1:

**Review by kidder — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]
squash and rebase :+1:

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**